scipy/optimize/tests/test__numdiff.py

```diff
@@ -0,0 +1,275 @@
+        jac_diff_f = approx_derivatives(self.fun_parametrized, x0,
+                                        method='forward', f0=f0, args=(c0, c1))
+        jac_diff_c = approx_derivatives(self.fun_parametrized, x0,
+                                        f0=f0, args=(c0, c1))
+        jac_true = self.jac_parametrized(x0, c0, c1)
+        assert_allclose(jac_diff_f, jac_true, rtol=1e-6)
+        assert_allclose(jac_diff_c, jac_true, rtol=1e-9)
+
+    def test_with_bounds_forward(self):
+        lb = -np.ones(2)
+        ub = np.ones(2)
+
+        x0 = np.array([-2.0, 0.2])
+        assert_raises(ValueError, approx_derivatives,
+                      self.fun_vector_vector, x0, bounds=(lb, ub))
+
+        x0 = np.array([-1.0, 1.0])
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0,
+                                      method='forward', bounds=(lb, ub))
+        jac_true = self.jac_vector_vector(x0)
+        assert_allclose(jac_diff, jac_true, rtol=1e-6)
+
+    def test_with_bounds_central(self):
+        lb = np.array([1.0, 1.0])
+        ub = np.array([2.0, 2.0])
+
+        x0 = np.array([1.0, 2.0])
+        jac_true = self.jac_vector_vector(x0)
+
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0)
+        assert_allclose(jac_diff, jac_true, rtol=1e-9)
+
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0,
+                                      bounds=(lb, None))
+        assert_allclose(jac_diff, jac_true, rtol=1e-9)
+
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0,
+                                      bounds=(None, ub))
+        assert_allclose(jac_diff, jac_true, rtol=1e-9)
+
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0,
+                                      bounds=(lb, ub))
+        assert_allclose(jac_diff, jac_true, rtol=1e-9)
+
+        lb = x0 - 1e-8
+        ub = x0 + 1e-6
+        jac_diff = approx_derivatives(self.fun_vector_vector, x0,
+                                      bounds=(lb, ub))
+        assert_allclose(jac_diff, jac_true, rtol=1e-6)
+
+    def test_check_derivatives(self):
+        x0 = np.array([-10.0, 10])
+        accuracy = check_derivatives(self.fun_vector_vector,
+                                     self.jac_vector_vector, x0)
+        assert_(accuracy < 1e-9)
+        accuracy = check_derivatives(self.fun_vector_vector,
+                                     self.jac_vector_vector, x0,
+                                     method='forward')
+        assert_(accuracy < 1e-6)
+
+        x0 = np.array([0.0, 0.0])
+        accuracy = check_derivatives(self.fun_zero_jacobian,
+                                     self.jac_zero_jacobian, x0)
+        assert_(accuracy == 0)
+        accuracy = check_derivatives(self.fun_zero_jacobian,
+                                     self.jac_zero_jacobian, x0,
+                                     method='forward')
+        assert_(accuracy == 0)
+
```

Comment:
And this also needs a test asserting that the difference scheme does switch to a one-sided one at the boundary. At a minimum, 

```
def f(x):
   x = np.atleast_1d(x)
   res = np.sin(x)
   res[x > np.pi] = np.nan
   return res 

deriv = approx_derivatives(f, x0=[np.pi], bounds=(None, np.pi))
assert_(np.isfinite(deriv).all())
```